C/services/common-plugin-interfaces/python/include/python_plugin_common_interface.h

```diff
@@ -0,0 +1,414 @@
+ */
+
+using namespace std;
+
+extern "C" {
+// This is the Python object initialised in:
+// South, Notification, Filter and North python plugin interfaces
+extern PyObject* pModule;
+
+// Common methods to all plugin interfaces
+static PLUGIN_INFORMATION *plugin_info_fn();
+static PLUGIN_HANDLE plugin_init_fn(ConfigCategory *);
+static void plugin_reconfigure_fn(PLUGIN_HANDLE*, const std::string&);
+static void plugin_shutdown_fn(PLUGIN_HANDLE);
+
+static void logErrorMessage();
+
+/**
+ * Destructor for PythonPluginHandle
+ *    - Free up owned references
+ *    - Unload python 3.5 interpreter
+ */
+static void PluginInterfaceCleanup()
+{
+	// Decrement pModule reference count
+	Py_CLEAR(pModule);
+
+	// Cleanup Python 3.5
+	Py_Finalize();
+}
+
+/**
+ * Returns function pointer that can be invoked to call 'plugin_info'
+ * function in python plugin
+ */
+static void* PluginInterfaceGetInfo()
+{
+	return (void *) plugin_info_fn;
+}
+
+/**
+ * Fill PLUGIN_INFORMATION structure from Python object
+ *
+ * @param pyRetVal      Python 3.5 Object (dict)
+ * @return              Pointer to a new PLUGIN_INFORMATION structure
+ *                              or NULL in case of errors
+ */
+static PLUGIN_INFORMATION *Py2C_PluginInfo(PyObject* pyRetVal)
+{
+	// Create returnable PLUGIN_INFORMATION structure
+	PLUGIN_INFORMATION *info = new PLUGIN_INFORMATION;
+
+	// these are borrowed references returned by PyDict_Next
+	PyObject *dKey, *dValue;
+	Py_ssize_t dPos = 0;
+
+	// dKey and dValue are borrowed references
+	while (PyDict_Next(pyRetVal, &dPos, &dKey, &dValue))
+	{
+		const char* ckey = PyUnicode_AsUTF8(dKey);
+		const char* cval = PyUnicode_AsUTF8(dValue);
+
+		char *valStr = new char [string(cval).length()+1];
+		std::strcpy (valStr, cval);
+
+		if(!strcmp(ckey, "name"))
+		{
+			info->name = valStr;
+		}
+		else if(!strcmp(ckey, "version"))
+		{
+			info->version = valStr;
+		}
+		else if(!strcmp(ckey, "mode"))
+		{
+			info->options = 0;
+			if (!strcmp(valStr, "async"))
+			{
+				info->options |= SP_ASYNC;
+			}
+			free(valStr);
+		}
+		else if(!strcmp(ckey, "type"))
+		{
+			info->type = valStr;
+		}
+		else if(!strcmp(ckey, "interface"))
+		{
+			info->interface = valStr;
+		}
+		else if(!strcmp(ckey, "config"))
+		{
+			info->config = valStr;
+		}
+	}
+
+	return info;
+}
+
+/**
+ * Function to invoke 'plugin_info' function in python plugin
+ */
+static PLUGIN_INFORMATION *plugin_info_fn()
+{
+	PyObject* pFunc; 
+	PyGILState_STATE state = PyGILState_Ensure();
+
+	// Fetch required method in loaded object
+	pFunc = PyObject_GetAttrString(pModule, "plugin_info");
+	if (!pModule || !pFunc)
+	{
+		Logger::getLogger()->info("plugin_handle: plugin_info(): pModule=%p, pFunc=%p",
+					   pModule,
+					   pFunc);
+	}
+
+	if (!pFunc || !PyCallable_Check(pFunc))
+	{
+		// Failure
+		if (PyErr_Occurred())
+		{       
+			logErrorMessage();
+		}
+
+		Logger::getLogger()->fatal("Cannot find method plugin_info in loaded python module");
```

Comment:
Can we give the module name in this error message?

Reply:
The Python module name is not available in these routines